wit/types.wit

```diff
@@ -1,44 +1,77 @@
-interface messaging-types {
+interface types {
     /// A connection to a message-exchange service (e.g., buffer, broker, etc.).
     resource client {
         connect: static func(name: string) -> result<client, error>;
     }
     
-    /// TODO(danbugs): This should be eventually extracted as an underlying type for other wasi-cloud-core interfaces.
-    resource error {
-        trace: static func() -> string;    
+    /// Errors that can occur when using the messaging interface.
+    variant error {
+        /// The requested option is not authorized. This could be a topic it doesn't have
+        /// permission to subscribe to, or a permission it doesn't have to perform a specific
+        /// action. This error is mainly used when calling `update-config`.
+        unauthorized,
+        /// The request or operation timed out.
+        timeout,
+        /// An error occurred with the connection. Includes a message for additional context
+        connection(string),
+        /// A catch all for other types of errors
+        other(string),
     }
   
-    /// There are two types of channels:
-    /// - publish-subscribe channel, which is a broadcast channel, and
-    /// - point-to-point channel, which is a unicast channel.
-    ///
-    /// The interface doesn't highlight this difference in the type itself as that's uniquely a consumer issue.
-    type channel = string;
+    /// The topic of a message. This is also called subject, channel, or group in various messaging
+    /// systems. What this value means is up to the implementation.
+    type topic = string;
   
-    /// Configuration includes a required list of channels the guest is subscribing to, and an optional list of extensions key-value pairs 
-    /// (e.g., partitions/offsets to read from in Kafka/EventHubs, QoS etc.).
-    record guest-configuration {
-        channels: list<channel>,
-        extensions: option<list<tuple<string, string>>>
+    /// Configuration includes a required list of channels the guest is subscribing to
+    record config {
+        channels: list<topic>,
     }
   
-    /// Format specification for messages 
-    ///  - more info: https://github.com/clemensv/spec/blob/registry-extensions/registry/spec.md#message-formats
-    ///  - message metadata can further decorate w/ things like format version, and so on.
-    enum format-spec {
-        cloudevents,
-        http,
-        amqp,
-        mqtt,
-        kafka,
-        raw
-    }
-  
-    /// A message with a binary payload, a format specification, and decorative metadata.
-    record message {
-        data: list<u8>,
-        format: format-spec,
-        metadata: option<list<tuple<string, string>>>
+    /// A message with a binary payload and additional information
+    resource message {
+        constructor(topic: topic, data: list<u8>, content-type: option<string>, metadata: option<list<tuple<string, string>>>);
```

Comment:
```suggestion
        constructor(topic: topic, data: list<u8>, opts: option<message-options>);
```

```wit
record message-options {
  content-type: option<string>
  metadata: option<list<tuple<string, string>>>
}
```

I feel it will be prudent to avoid breaking changes in the function signature

Reply:
@lukewagner I'm curious on your thoughts here on how constructors should be handled. In particular:

1. Is it idiomatic to have another record type you use in a constructor?
2. Are there any other considerations around breaking a constructor API that we should take into account here? At this point I don't think the `@since` and `@unstable` would necessarily help here

Reply:
To carve out a way to add optional fields in a semver-compatible way (at least in a 0.2 timeframe before we relax subtyping), you can use a resource type with methods for each field, which is what [wasi-http does with `request-options`](https://github.com/WebAssembly/wasi-http/blob/main/wit/types.wit#L331).